# kellets_whelk_rnaseq

## Project description
This is a project analyzing population structure and genetic adaptation of Kellet's whelks (Kelletia kelletii)  --- a range expanding marine species with high gene flow potential. We performed a series of crosses on Kellet's whelks collected from its historical and recently colonized range, and conducted RNA-Seq on offspring that we reared in a common garden environment. We conducted differential expression analysis on 54 samples with historical-range and expanded-range ancestry. We also used transcriptome derived SNPs to examine population structure and genetic adaptation. 

Corresponding article [place holder for citation]

## Scripts
Code for project is in [scripts](https://github.com/ChristieLab/kellets_whelk_rnaseq/tree/main/scripts). Analyses are split up into directories: 
- **1. Study system.** R code associated with figure 1. 
- **2. Sequence processing.** Process RNA transcripts and conduct DEG analysis 
- **3. SNP calling.** Contains for SNP variant calling using ANGSD 
- **4. Population structure and population assignment.** R codes analyszing transcriptome-derived SNPs using snpR (Figure 2) 
- **5. Candidate loci driving differntiation.**  R codes analyszing transcriptome-derived SNPs using snpR, population assignment using rubias and ranger (Figure 3). R code analyzing the TPI gene (Figure 4)
- **6. Gene Ontology and WGCNA.** R codes associated with analyses of expression data (Supplemental)  
  

## Data
Data files are in data [place holder for link] and dryad [place holder for link] 
